cmd/cat/cat.go

```diff
@@ -11,126 +10,65 @@ import (
 )
 
-type conf struct {
-	offset             int64
-	recordNum          int
-	recordCount        int
-	fileName           string
-	filter             *filter.Filter
-	showWarcHeader     bool
-	showProtocolHeader bool
-	showPayload        bool
+type catConfig struct {
+	Offset             int64
+	RecordNum          int
+	RecordCount        int
+	FileName           string
+	Filter             *filter.Filter
+	ShowWarcHeader     bool
+	ShowProtocolHeader bool
+	ShowPayload        bool
 }
 
-func NewCommand() *cobra.Command {
-	var cmd = &cobra.Command{
-		Use:   "cat",
-		Short: "Concatenate and print warc files",
-		Long:  ``,
-		Example: `# Print all content from a WARC file
-warc cat file1.warc.gz
-
-# Pipe payload from record #4 into the image viewer feh
-warc cat -n4 -P file1.warc.gz | feh -`,
-		RunE: parseArgumentsAndCallCat,
-	}
-
-	cmd.Flags().Int64P(flag.Offset, "o", -1, flag.OffsetHelp)
-	cmd.Flags().IntP(flag.RecordNum, "n", -1, flag.RecordNumHelp)
-	cmd.Flags().IntP(flag.RecordCount, "c", 0, flag.RecordCountHelp+" Defaults to show all records except if -o or -n option is set, then default is one.")
-	cmd.Flags().BoolP(flag.ShowWarcHeader, "w", false, flag.ShowWarcHeaderHelp)
-	cmd.Flags().BoolP(flag.ShowProtocolHeader, "p", false, flag.ShowProtocolHeaderHelp)
-	cmd.Flags().BoolP(flag.ShowPayload, "P", false, flag.ShowPayloadHelp)
-	cmd.Flags().StringArray(flag.RecordId, []string{}, flag.RecordIdHelp)
-	cmd.Flags().StringSliceP(flag.RecordType, "t", []string{}, flag.RecordTypeHelp)
-	cmd.Flags().StringP(flag.ResponseCode, "S", "", flag.ResponseCodeHelp)
-	cmd.Flags().StringSliceP(flag.MimeType, "m", []string{}, flag.MimeTypeHelp)
-
-	return cmd
-}
-
-func parseArgumentsAndCallCat(cmd *cobra.Command, args []string) error {
-	config := &conf{}
-	if len(args) == 0 {
-		return errors.New("missing file name")
-	}
-	config.fileName = args[0]
-	config.offset = viper.GetInt64(flag.Offset)
-	config.recordCount = viper.GetInt(flag.RecordCount)
-	config.recordNum = viper.GetInt(flag.RecordNum)
-	config.showWarcHeader = viper.GetBool(flag.ShowWarcHeader)
-	config.showProtocolHeader = viper.GetBool(flag.ShowProtocolHeader)
-	config.showPayload = viper.GetBool(flag.ShowPayload)
-
-	if (config.offset >= 0 || config.recordNum >= 0) && config.recordCount == 0 {
-		config.recordCount = 1
-	}
-	if config.offset < 0 {
-		config.offset = 0
-	}
-
-	config.filter = filter.NewFromViper()
-
-	if !(config.showWarcHeader || config.showProtocolHeader || config.showPayload) {
-		config.showWarcHeader = true
-		config.showProtocolHeader = true
-		config.showPayload = true
-	}
-	return runE(config)
-}
-
-func runE(c *conf) error {
-	readFile(c, c.fileName)
-	return nil
-}
-
-func readFile(c *conf, fileName string) {
-	wf, err := gowarc.NewWarcFileReader(fileName, c.offset, gowarc.WithBufferTmpDir(viper.GetString(flag.TmpDir)))
-	defer func() { _ = wf.Close() }()
+func listRecords(config *catConfig, fileName string) ([]gowarc.WarcRecord, error) {
```

Comment:
I understand changing the signature of the function would make it testable but it would potentially blow up the memory consumption when cat'ing one or more large WARC files.

I would suggest using a send channel as a parameter 

```golang
func listRecords(config *catConfig, fileName string, records chan<- string) error {
  defer close(records)
  warcFileReader, err := gowarc.NewWarcFileReader(fileName, config.Offset, gowarc.WithBufferTmpDir(viper.GetString(flag.TmpDir)))
  defer func() { _ = warcFileReader.Close() }()
  ...
  go func() {
    warcRecord, _, _, err := warcFileReader.Next()
    ...
    records <- warcRecord
    ...
    // if any errors during iteration we print them
   fmt.Fprintf(os.Stderr, ..:)
    ...
 }()
 return nil
}
```
then the caller can decide the buffering and perform the actual output:

```golang

bufSize := 100 // we can live with eg. 100 records in memory
records := make(chan gowarc.WarcRecord, bufSize?)

err := listRecords(config, filename, records)
if err != nil {
   // handle error from gowerc.NewWarcFileReader
}

// iterate over the record from the records channel
for record := range records {
  // print the record according to config
  ...
   // and a test can append it to a slice here 
}
```

Reply:
I see your point, but I would also argue that trying to ensure that this project runs with a specific performance requirements, without actually having any tests for those requirements is a recipe for failure. If we are to enforce specific memory footprint requirements without any testing what-so-ever, how do we even know if we fulfill those requirements today?

For example, what is a "large WARC file"? Kilobytes sized ones? GB? TB?
If this project is only ever going to be a CLI tool, does it matter that it uses more memory that it should?

Anyway, since memory consumption is, apparently, important for this tool, I quickly profiled the code.
The important bits, before:
![image](https://github.com/nlnwa/warchaeology/assets/22957880/7115412c-2e52-4bf9-984e-c95c2b30725f)

After:
![image](https://github.com/nlnwa/warchaeology/assets/22957880/57baf516-c948-4a22-b8e6-999b2a039f69)

So you are correct that the memory footprint is different, and because of its importance I will remove that latest commit so the other two commits are no longer blocked by this PR.

The TL:DR is that it is difficult to enforce a certain memory footprint without benchmark test or automatic profiling.